### PR TITLE
use replica storage for status calls

### PR DESCRIPTION
**Pull request by mkelly121**
Added a read-only storage engine to be used in job status calls.

New methods `FetchAsync`, `GetJobReportAsync` and `GetJobRequestAsync` will use this read-only storage

### CODE REVIEW

File: cmds/contest/main.go [hunk 1]
```diff
@@ -44,22 +44,43 @@ func main() {
...
 	if err != nil {
 		log.Warningf("Could not determine storage version: %v", err)
 	} else {
 		log.Infof("Storage version: %d", dbVer)
 	}
 
+	// replica storage initialization
+	log.Infof("Using database URI for replica storage: %s", *flagDBURI)
+
+	replicaDBURI := *flagDBURI
```
Comment (tfg13) [inline cmds/contest/main.go:69]: Drop a comment here that this is just pointing to the main database for now, but could be used to connect to a replica

File: cmds/contest/main.go [hunk 2]
```diff
@@ -44,22 +44,43 @@ func main() {
...
+	if err := storage.SetStorageAsync(s); err != nil {
+		log.Fatalf("Could not set replica storage: %v", err)
+	}
+
+	dbVer, err = s.Version()
+	if err != nil {
+		log.Warningf("Could not determine storage version: %v", err)
+	} else {
+		log.Infof("Storage version: %d", dbVer)
+	}
```
Comment (tfg13) [inline cmds/contest/main.go:83]: lets fetch the version and compare it to the main one. This is by no means a check that we actually talk to the same DB, but if the version numbers are different something is very wrong and we should exit immediately

File: pkg/jobmanager/jobmanager.go [hunk 3]
```diff
@@ -60,7 +60,7 @@ type JobManager struct {
 	jobsMu sync.Mutex
 	jobsWg sync.WaitGroup
 
-	jobStorageManager storage.JobStorage
+	jobStorageManager storage.JobStorageManager
```
Comment (tfg13) [inline pkg/jobmanager/jobmanager.go:63]: accidental change?

File: pkg/storage/storage.go [hunk 4]
```diff
@@ -56,3 +57,29 @@ func SetStorage(storageEngine Storage) error {
...
+	if storage == nil {
+		return nil, fmt.Errorf("no storage engine assigned")
+	}
+	return storage, nil
+}
+
+// SetStorage sets the desired storage engine for read-only events. Switching to a new
+// storage engine implies garbage collecting the old one, with possible loss of
+// pending events if not flushed correctly
+func SetStorageAsync(storageEngine Storage) error {
```
Comment (tfg13) [inline pkg/storage/storage.go:72]: Maybe name this `SetAsyncStorage`, also copy-paste in first word of the doc comment :)

**Review by tfg13 — CHANGES_REQUESTED** [review]
This looks good! Just some minor changes left

File: pkg/storage/job.go [hunk 5]
```diff
@@ -48,19 +52,22 @@ func (jsm JobStorageManager) StoreJobReport(report *job.JobReport) error {
...
 
-// GetJobReport fetches a job report to the storage layer
+// GetJobReport fetches a job report from the storage layer
 func (jsm JobStorageManager) GetJobReport(jobID types.JobID) (*job.JobReport, error) {
-	return storage.GetJobReport(jobID)
+	report, err := storage.GetJobReport(jobID)
+	if err != nil {
+		return nil, err
+	}
+	return report, nil
```
Comment (tfg13) [inline pkg/storage/job.go]: isn't this exactly the same as:
```suggestion
	return storage.GetJobReport(jobID)
```

File: pkg/storage/job.go [hunk 6]
```diff
@@ -48,19 +52,22 @@ func (jsm JobStorageManager) StoreJobReport(report *job.JobReport) error {
...
-// GetJobReport fetches a job report from the storage layer
+// GetJobReportAsync fetches a job report from the read-only storage layer
 func (jsm JobStorageManager) GetJobReportAsync(jobID types.JobID) (*job.JobReport, error) {
-	return storageAsync.GetJobReport(jobID)
-}
-
-// ListJobs returns list of job IDs matching the query
-func (jsm JobStorageManager) ListJobs(query *JobQuery) ([]types.JobID, error) {
-	return storage.ListJobs(query)
+	report, err := storageAsync.GetJobReport(jobID)
```
Comment (tfg13) [inline pkg/storage/job.go:66]: same

**Review by tfg13 — APPROVED** [review]
one super minor nit, only if you want to

**Comment by rojer9-fb** [comment]
just ran into this while rebasing - no objections in principle but the "Async" suffix for storage methods seems confusing - usually "FooAsync" means "do Foo in a non-blocking manner", not "obtain a possibly stale version of Foo".

**Comment by tfg13** [comment]
Feel free to rename it, this was the best I could come up with on short notice, not married to the name :)

**Pull request merged**